### ISSUE TITLE
dvc cannot handle gitignore patterns

### ISSUE THREAD

**Issue by muaali** [iterative/dvc#5748]
**Problem Description:**

I am gitignoring most subdirectories of my `dataset` folder and trying to dvc track few files from my `dataset` subdirectories. Even after adding those files to dvc, it seems that dvc cannot track them. Local cache is built and they are hashed, but not tracked by dvc apparently.

**Error output:**
After adding `.dvc` files with `dvc add`, and running `dvc push` gives the following output:
```
There are no data or pipelines tracked in this project yet.           
See <https://dvc.org/doc/start> to get started!
```

**Debugging:**

```
dvc status
dvc status path/to/file.dvc
dvc status -R .
```
give output 
```
There are no data or pipelines tracked in this project yet.           
See <https://dvc.org/doc/start> to get started!
```

**Reproduce**

Create a folder structure as shown in the gitignore file below. Add `gitignore` to the main directory with the following content :
```
!/package_root/dataset/
/package_root/dataset/*
!/package_root/dataset/train_dataset/
/package_root/dataset/train_dataset/*
!/package_root/dataset/train_dataset/labels/
/package_root/dataset/train_dataset/labels/*
!/package_root/dataset/train_dataset/labels/*.dvc
!/package_root/dataset/train_dataset/Images/
/package_root/dataset/train_dataset/Images/*
!/package_root/dataset/train_dataset/Images/*.dvc

```
Run following commands: 

```
dvc add package_root/dataset/train_dataset/labels/labels.csv #labels.csv is a dummy csv file.
git add package_root/dataset/train_dataset/labels/labels.csv.dvc
dvc add package_root/dataset/train_dataset/Images/* #contain images, can be any image.
git add package_root/dataset/train_dataset/Images/*.dvc
dvc push
```
gives the output : 
```
There are no data or pipelines tracked in this project yet.           
See <https://dvc.org/doc/start> to get started!
```
And with more verbose output : 
`dvc push -v`
gives output
```
2021-03-31 16:18:23,019 DEBUG: Check for update is enabled.
2021-03-31 16:18:23,134 DEBUG: Preparing to upload data to 's3://<bucket_name>/<directory>'
2021-03-31 16:18:23,134 DEBUG: Preparing to collect status from s3://<bucket_name>/<directory>
2021-03-31 16:18:23,134 DEBUG: Collecting information from local cache...
2021-03-31 16:18:23,134 DEBUG: Collecting information from remote cache...                                                                                 
Everything is up to date.                                                                                                                                  
2021-03-31 16:18:23,137 DEBUG: <s3-profile> is enabled.
2021-03-31 16:18:23,169 DEBUG: Trying to spawn '['daemon', '-q', '<s3-profile>', '/tmp/tmp81bxtd8c']'
2021-03-31 16:18:23,170 DEBUG: Spawned '['daemon', '-q', '<s3-profile>', '/tmp/tmp81bxtd8c']'

```

discord context: https://discord.com/channels/485586884165107732/485596304961962003/826748346882195486

**Comment by pared** [comment]
Related: https://github.com/iterative/dvc/issues/5650

**Comment by pmrowla** [comment]
This is the edge case described in the last example from the `gitignore` docs: https://git-scm.com/docs/gitignore#_examples

The include/exclude patterns are not being merged properly when used in multiple levels of nesting, will need to be fixed somewhere in dulwich's `IgnoreFilter`:

https://github.com/dulwich/dulwich/blob/76dd8356f9d8870f44226e4232c24fbc1e573e93/dulwich/ignore.py#L202

**Comment by dberenbaum** [comment]
Is there a Dulwich issue or PR for this?

**Comment by pmrowla** [comment]
@dberenbaum no, someone still has to do the actual investigation on the dulwich side to see what exactly needs to be fixed. Someone should also test using libgit/pygit's gitignore implementation as well, because if this edge case is already properly supported there it may be easier for us to just switch to that rather than fixing this particular issue in dulwich.

**Comment by karajan1001** [comment]
I'm quite familiar with git-ignore patterns, I can help them if needed.

**Comment by tpietruszka** [comment]
Experiencing a variant of this problem as well - in fact it's preventing all my teammates and me from upgrading to dvc v2.

We use a pattern where we `dvc import` (and `dvc add`) files in subdirectories of the `data` directory

Then our `.gitignore` looks like:
```
data/**
!data/**/
!data/**/*.dvc
```
Git handles it well, DVC does not (since v2).
After upgrading to DVC v2, `dvc pull` does not actually pull anything.

Toy example: https://github.com/tpietruszka/dvc-import-bug

**Comment by tpietruszka** [comment]
Actually, I'm not sure this is about `dulwich`.

Tried using `dulwich` directly, to browse the repo mentioned above, which causes problems in DVC.

Result:
```
In [1]: from dulwich.repo import Repo

In [2]: r = Repo(".")

In [3]: index = r.open_index()

In [4]: list(index)
Out[4]:
[b'.dvc/.gitignore',
 b'.dvc/config',
 b'.dvc/plots/confusion.json',
 b'.dvc/plots/confusion_normalized.json',
 b'.dvc/plots/default.json',
 b'.dvc/plots/linear.json',
 b'.dvc/plots/scatter.json',
 b'.dvc/plots/smooth.json',
 b'.dvcignore',
 b'.gitignore',
 b'README.md',
 b'data/import/owl_sticker.png.dvc']
```

So `dulwich` does know that the file of interest (`data/import/owl_sticker.png.dvc`) is tracked.

**Comment by tpietruszka** [comment]
Something strange happens around here: https://github.com/iterative/dvc/blob/4e792ae61c5927ab2e5f6a6914d985d43aa705b4/dvc/repo/stage.py#L468-L470

When I set a breakpoint in line 467 and displayed the results of `repo.dvcignore.walk()` - the file of interest was there:
```
python -m pdb -m dvc pull
> /home/tomasz/.local/lib/python3.8/site-packages/dvc/__main__.py(1)<module>()
-> """Main entry point for DVC command line tool."""
(Pdb) b dvc/repo/stage.py:467
Breakpoint 1 at /home/tomasz/.local/lib/python3.8/site-packages/dvc/repo/stage.py:467
(Pdb) c
> /home/tomasz/.local/lib/python3.8/site-packages/dvc/repo/stage.py(467)collect_repo()
-> stages = []
(Pdb) p list(self.repo.dvcignore.walk(self.fs, self.repo.root_dir))
[('/home/tomasz/dvc-import-test', ['data', 'local'], ['.dvcignore', '.gitignore', 'README.md']), ('/home/tomasz/dvc-import-test/data', ['import'], []), ('/home/tomasz/dvc-import-test/data/import', [], ['owl_sticker.png.dvc']), ('/home/tomasz/dvc-import-test/local', [], [])]
```

However, if I then set a breakpoint in line 471, to see each iteration of the loop - it only gets triggered twice, with the first 2 tuples visible above - out of 4 tuples total!

I would bet that's because of these 2 bits of logic:
https://github.com/iterative/dvc/blob/4e792ae61c5927ab2e5f6a6914d985d43aa705b4/dvc/repo/stage.py#L489
https://github.com/iterative/dvc/blob/4e792ae61c5927ab2e5f6a6914d985d43aa705b4/dvc/ignore.py#L275-L277

Basically, the client of the `walk` generator is modifying the `dirs` list in place, while the generator is using the same list to... generate. I don't understand the purpose of it and it's really hard to reason about. I wouldn't be surprised if it was causing unexpected behavior in some cases.


Is this at all useful @pmrowla? Happy to keep digging, but I don't really get the full picture here, so stopping for now and waiting for feedback 😉

**Comment by tpietruszka** [comment]
OK, I think I got to the bottom of this!

When we call `is_ignored(dir_path)` (https://github.com/iterative/dvc/blob/4e792ae61c5927ab2e5f6a6914d985d43aa705b4/dvc/repo/stage.py#L463-L465) we don't include a trailing slash in the `dir_path`.

But we should, according to a `dulwich` docstring https://github.com/dulwich/dulwich/blob/76dd8356f9d8870f44226e4232c24fbc1e573e93/dulwich/ignore.py#L228-L231
where these paths end up


Working on a PR

**Comment by efiop** [comment]
Thank you @tpietruszka ! :pray: Releasing new version right now.
